src/iceberg/avro/avro_direct_encoder.cc

```diff
@@ -0,0 +1,417 @@
+      const auto& binary_array =
+          internal::checked_cast<const ::arrow::BinaryArray&>(array);
+      std::string_view value = binary_array.GetView(row_index);
+      ctx->bytes_scratch.assign(value.begin(), value.end());
+      encoder.encodeBytes(ctx->bytes_scratch);
+      return {};
+    }
+
+    case ::avro::AVRO_FIXED: {
+      // Handle UUID
+      if (avro_node->logicalType().type() == ::avro::LogicalType::UUID) {
+        const auto& extension_array =
+            internal::checked_cast<const ::arrow::ExtensionArray&>(array);
+        const auto& fixed_array =
+            internal::checked_cast<const ::arrow::FixedSizeBinaryArray&>(
+                *extension_array.storage());
+        std::string_view value = fixed_array.GetView(row_index);
+        ctx->bytes_scratch.assign(value.begin(), value.end());
+        encoder.encodeFixed(ctx->bytes_scratch.data(), ctx->bytes_scratch.size());
+        return {};
+      }
+
+      // Handle DECIMAL
+      if (avro_node->logicalType().type() == ::avro::LogicalType::DECIMAL) {
+        const auto& decimal_array =
+            internal::checked_cast<const ::arrow::Decimal128Array&>(array);
+        std::string_view decimal_value = decimal_array.GetView(row_index);
+        ctx->bytes_scratch.assign(decimal_value.begin(), decimal_value.end());
+        // Arrow Decimal128 bytes are in little-endian order, Avro requires big-endian
+        std::ranges::reverse(ctx->bytes_scratch);
+        encoder.encodeFixed(ctx->bytes_scratch.data(), ctx->bytes_scratch.size());
+        return {};
+      }
+
+      // Handle regular FIXED
+      const auto& fixed_array =
+          internal::checked_cast<const ::arrow::FixedSizeBinaryArray&>(array);
+      std::string_view value = fixed_array.GetView(row_index);
+      ctx->bytes_scratch.assign(value.begin(), value.end());
+      encoder.encodeFixed(ctx->bytes_scratch.data(), ctx->bytes_scratch.size());
+      return {};
+    }
+
+    case ::avro::AVRO_RECORD: {
+      if (array.type()->id() != ::arrow::Type::STRUCT) {
+        return InvalidArgument("AVRO_RECORD expects StructArray, got {}",
+                               array.type()->ToString());
+      }
+      if (!type.is_nested()) {
+        return InvalidArgument("AVRO_RECORD expects nested type, got type {}",
+                               type.ToString());
+      }
+
+      const auto& struct_array =
+          internal::checked_cast<const ::arrow::StructArray&>(array);
+
+      // AVRO_RECORD corresponds to Iceberg StructType (including Schema which extends
+      // StructType). Note: ListType and MapType are encoded as AVRO_ARRAY and AVRO_MAP
+      // respectively, not AVRO_RECORD.
+      if (type.type_id() != TypeId::kStruct) {
+        return InvalidArgument("AVRO_RECORD expects StructType, got type {}",
+                               type.ToString());
+      }
+
+      // Safe cast: type_id() == kStruct guarantees this is StructType or Schema
+      // (Schema extends StructType)
```

Comment:
```suggestion
      if (type.type_id() != TypeId::kStruct) {
        return InvalidArgument("AVRO_RECORD expects StructType, got type {}",
                               type.ToString());
      }

```

I don't think we need these obvious comments.